File: LeetCode/HashMap/JavaScript/happyNumber.js
// Write an algorithm to determine if a number n is happy.

// A happy number is a number defined by the following process:

// Starting with any positive integer, replace the number by the sum of the squares of its digits.
// Repeat the process until the number equals 1 (where it will stay), or it loops endlessly in a cycle which does not include 1.
// Those numbers for which this process ends in 1 are happy.
// Return true if n is a happy number, and false if not.

// Example 1:

// Input: n = 19
// Output: true
// Explanation:
// 12 + 92 = 82
// 82 + 22 = 68
// 62 + 82 = 100
// 12 + 02 + 02 = 1
// Example 2:

// Input: n = 2
// Output: false

// Constraints:

// 1 <= n <= 231 - 1

// TC: O(k*r) where k is the number of values computed as sum of square of digits until we find 1 or a duplicate. It could go to the limit of an integer.
//            r is the number of digits per iteration to compute the sum of squares.
// SC: O(k) where k is the number of values computed as sum of square of digits until we find 1 or a duplicate. It could go to the limit of an integer

/**
 * @param {number} n
 * @return {boolean}
 */
var isHappy = function (n) {
  // We'll use a set to store the values that we've already visited
  // This is used to detect if we've reached a loop
  let visited = new Set();

  while (!visited.has(n)) {
    // add n to our visited list
    visited.add(n);

    // compute the new n which is the square of each digits of the previous n
    n = computeN(n);
    // console.log(n_)

    // check if it's a happy number
    if (n === 1) {
      return true;
    }
  }

  // This is our helper function
  function computeN(n) {
    let res = 0;
    // This goes through all the digit of n and accumulate the square of all the digits
    while (n) {
      let temp = n % 10; // This returns the most significant digit
      res += temp ** 2; // let's square the most significant digit and accumulate it on the res variable
      n = (n / 10) | 0; // we want to perform an integer division so we can get the next digit
      // console.log(n_)
    }

    return res;
  }

  return false;
};

File: LeetCode/HashMap/python/happyNumber.py
from typing import Dict


class Solution:
    def isHappy(self, n: int) -> bool:
        visited = set()

        while n not in visited:
            # print ("before" + str(visited))
            visited.add(n)
            # print ("after" + str(visited))            

            # Compute the new n
            n = self.computeN(n)

            if n == 1:
                return True

        return False
    
    # This goes through all the digit of n and accumulate the square of all the digits
    def computeN(self, n: int) -> int:
        res = 0

        while n > 0 :
            temp = n % 10 # Get the most significant digit (MSD)
            res += (temp**2) # Compute the square of the MSD and accumulate to the res variable
            n = (n//10) # Get the next digit of n
        
        return res

    @staticmethod
    def testSolution(input: Dict[str, int | bool]) -> None:
        print("Input: n: " + str(input["n"]))
        print("Expected: " + str(input["expected"]))
        res = Solution().isHappy(input["n"])
        print("Result: " + str(res))
        print("PASS" if res == input["expected"] else "FAIL")
        print("-" * 50)
        

if __name__ == "__main__":
    input = {"n": 19, "expected": True}
    Solution.testSolution(input)

    input = {"n": 2, "expected": False}
    Solution.testSolution(input)